### ISSUE TITLE
travis build stops for a reason

### ISSUE THREAD

**Issue by miekg** [coredns/coredns#1724]
Def. seeing this a lot more:

~~~
=== RUN   TestNameFromRight
--- PASS: TestNameFromRight (0.00s)
PASS
2018/04/24 06:49:11 [ERROR] plugin/file: Failed to open zone "miek.nl." in "go-test-tmpfile739461045": open go-test-tmpfile739461045: no such file or directory
2018/04/24 06:49:11 [ERROR] plugin/file: Failed to open zone "miek.nl." in "go-test-tmpfile739461045": open go-test-tmpfile739461045: no such file or directory
ok  	github.com/coredns/coredns/plugin/file	3.400s
=== RUN   TestLess
--- PASS: TestLess (0.00s)
PASS
ok  	github.com/coredns/coredns/plugin/file/tree	1.012s
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated


=== RUN   TestNameFromRight
--- PASS: TestNameFromRight (0.00s)
PASS
2018/04/24 06:49:11 [ERROR] plugin/file: Failed to open zone "miek.nl." in "go-test-tmpfile739461045": open go-test-tmpfile739461045: no such file or directory
2018/04/24 06:49:11 [ERROR] plugin/file: Failed to open zone "miek.nl." in "go-test-tmpfile739461045": open go-test-tmpfile739461045: no such file or directory
ok  	github.com/coredns/coredns/plugin/file	3.400s
=== RUN   TestLess
--- PASS: TestLess (0.00s)
PASS
ok  	github.com/coredns/coredns/plugin/file/tree	1.012s
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated  
~~~

Haven't look in detail, gut feeling says it started with the merge of #1709

**Comment by miekg** [comment]
ha! Saw it in the act forward.test takes a longish time

**Comment by miekg** [comment]
stracing a `go test -v` in plugin/forward:
~~~
miek     27089  0.0  0.6 569156 25852 pts/2    Sl+  08:29   0:00 /tmp/go-build316954149/b001/forward.test -test.v=true
miek     27285  0.0  0.0  12788   936 pts/3    S+   08:32   0:00 grep test
% strace -p 27089 
strace: Process 27089 attached
epoll_pwait(4, 
~~~~

**Comment by miekg** [comment]
suspecting c9b09d04
[fwdebug c9b09d04] Revert "plugin/forward: gracefull stop (#1701)"

**Comment by miekg** [comment]
`% for i in {0..10}; do go t -v; done`; went to completion without a hitch.
Now doing `{0...100}`.

If we want graceful stop (and we want it) we should a) check proxy and b) prolly extract out the bits in a plugin/pkg/graceful if possible for easier testing.

**Comment by miekg** [comment]
Still happening on master:

-- PASS: TestNameFromRight (0.00s)
PASS
2018/04/25 08:19:37 [ERROR] plugin/file: Failed to open zone "miek.nl." in "go-test-tmpfile945862021": open go-test-tmpfile945862021: no such file or directory
2018/04/25 08:19:38 [ERROR] plugin/file: Failed to open zone "miek.nl." in "go-test-tmpfile945862021": open go-test-tmpfile945862021: no such file or directory
ok  	github.com/coredns/coredns/plugin/file	3.378s
=== RUN   TestLess
--- PASS: TestLess (0.00s)
PASS
ok  	github.com/coredns/coredns/plugin/file/tree	1.014s
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated

**Comment by miekg** [comment]
I'm closing all channels in a local test, doesn't hang; if I output anything to the screen it doesn't hang anymore, making debugging hard :/

**Comment by rdrozhdzh** [comment]
The last failure happened after we merged #1727, right? And it didn't panic, so seems this is something different than writing to abandoned channel. Btw, I cannot see failed jobs in travis, did you re-run them?

I looked through the code again, and still cannot figure out what can be wrong. Currently I tend to re-write the test and replace WaitGroup with events sent by channels. This way we could also add timer channel and panic (with stack trace) if some goroutine has not finished say in 10 seconds

Having stack trace, we could better understand what causes the lock

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] travis build ..." ]
>The last failure happened after we merged #1727, right? And it didn't panic, so seems this is something different than writing to abandoned channel. Btw, I cannot see failed jobs in travis, did you re-run them?

I didn't merge #1725.... we could revert and add proxy_test.go back in; that we 
(maybe) simplify bits. I think I did see it fail then too. I can def. replicate 
it locally by running enough tests.

yes, I re-ran the travis jobs as these failures were not related to the PR.

>I looked through the code again, and still cannot figure out what can be wrong. Currently I tend to re-write the test and replace WaitGroup with events sent by channels. This way we could also add timer channel and panic (with stack trace) if some goroutine has not finished say in 10 seconds
>
>Having stack trace, we could better understand what causes the lock

Indeed! This is like a good plan.

**Comment by miekg** [comment]
If I do revert that commit again and add some tests, I get.
Maybe reverting is still an option?
~~~
==================
WARNING: DATA RACE
Read at 0x00c4202349c0 by goroutine 88:
  runtime.chansend()
      /home/miek/upstream/go/src/runtime/chan.go:140 +0x0
  github.com/coredns/coredns/plugin/forward.(*Proxy).connect()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/persistent.go:135 +0x148
  github.com/coredns/coredns/plugin/forward.TestProxyClose.func5()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/proxy_test.go:48 +0x9e

Previous write at 0x00c4202349c0 by goroutine 25:
  runtime.closechan()
      /home/miek/upstream/go/src/runtime/chan.go:325 +0x0
  github.com/coredns/coredns/plugin/forward.newTransport.func1()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/persistent.go:52 +0xa8

Goroutine 88 (running) created at:
  github.com/coredns/coredns/plugin/forward.TestProxyClose()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/proxy_test.go:47 +0x3b4
  testing.tRunner()
      /home/miek/upstream/go/src/testing/testing.go:777 +0x16d

Goroutine 25 (finished) created at:
  github.com/coredns/coredns/plugin/forward.newTransport()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/persistent.go:47 +0x273
  github.com/coredns/coredns/plugin/forward.NewProxy()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/proxy.go:35 +0x144
  github.com/coredns/coredns/plugin/forward.TestProxyClose()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/proxy_test.go:30 +0x195
  testing.tRunner()
      /home/miek/upstream/go/src/testing/testing.go:777 +0x16d
==================
==================
WARNING: DATA RACE
Read at 0x00c4202349c0 by goroutine 40:
  runtime.chansend()
      /home/miek/upstream/go/src/runtime/chan.go:140 +0x0
  github.com/coredns/coredns/plugin/forward.(*Proxy).connect()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/persistent.go:135 +0x148
  github.com/coredns/coredns/plugin/forward.TestProxyClose.func6()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/proxy_test.go:52 +0x9e

Previous write at 0x00c4202349c0 by goroutine 25:
  runtime.closechan()
      /home/miek/upstream/go/src/runtime/chan.go:325 +0x0
  github.com/coredns/coredns/plugin/forward.newTransport.func1()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/persistent.go:52 +0xa8

Goroutine 40 (running) created at:
  github.com/coredns/coredns/plugin/forward.TestProxyClose()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/proxy_test.go:51 +0x432
  testing.tRunner()
      /home/miek/upstream/go/src/testing/testing.go:777 +0x16d

Goroutine 25 (finished) created at:
  github.com/coredns/coredns/plugin/forward.newTransport()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/persistent.go:47 +0x273
  github.com/coredns/coredns/plugin/forward.NewProxy()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/proxy.go:35 +0x144
  github.com/coredns/coredns/plugin/forward.TestProxyClose()
      /home/miek/g/src/github.com/coredns/coredns/plugin/forward/proxy_test.go:30 +0x195
  testing.tRunner()
      /home/miek/upstream/go/src/testing/testing.go:777 +0x16d
==================
panic: send on closed channel

goroutine 95 [running]:
github.com/coredns/coredns/plugin/forward.(*transport).Dial(...)
	/home/miek/g/src/github.com/coredns/coredns/plugin/forward/persistent.go:135
github.com/coredns/coredns/plugin/forward.(*Proxy).Dial(...)
	/home/miek/g/src/github.com/coredns/coredns/plugin/forward/proxy.go:67
github.com/coredns/coredns/plugin/forward.(*Proxy).connect(0xc42018c500, 0xe2db20, 0xc420098058, 0xc420232b40, 0xe322c0, 0x11df158, 0x0, 0x0, 0x0, 0x0, ...)
	/home/miek/g/src/github.com/coredns/coredns/plugin/forward/connect.go:44 +0x149
github.com/coredns/coredns/plugin/forward.TestProxyClose.func5(0xc42018c500, 0xe2db20, 0xc420098058, 0xc420232b40, 0xe322c0, 0x11df158, 0x0, 0x0, 0x0, 0x0, ...)
	/home/miek/g/src/github.com/coredns/coredns/plugin/forward/proxy_test.go:48 +0x9f
created by github.com/coredns/coredns/plugin/forward.TestProxyClose
	/home/miek/g/src/github.com/coredns/coredns/plugin/forward/proxy_test.go:47 +0x3b5
exit status 2
FAIL	github.com/coredns/coredns/plugin/forward	23.064s
~~~

**Comment by miekg** [comment]
I think i have a PR

**Comment by miekg** [comment]
https://github.com/coredns/coredns/pull/1734

**Comment by rdrozhdzh** [comment]
I'm on vac this week, and currently have no configured git environment :(
The reworked test which I mentioned before should look like
```
	for i := 0; i < 100; i++ {
		p := NewProxy(s.Addr, nil /* no TLS */)
		p.start(hcDuration)

		doneCnt := 0
		doneCh := make(chan bool)
		timeCh := time.After(10 * time.Second)
		go func() {
			p.connect(ctx, state, false, false)
			doneCh <- true
		}()
		go func() {
			p.connect(ctx, state, true, false)
			doneCh <- true
		}()
		go func() {
			p.close()
			doneCh <- true
		}()
		go func() {
			p.connect(ctx, state, false, false)
			doneCh <- true
		}()
		go func() {
			p.connect(ctx, state, true, false)
			doneCh <- true
		}()

		for doneCnt < 5 {
			select {
			case <-doneCh:
				doneCnt++
			case <-timeCh:
				t.Error("TestProxyClose is running too long, dumping goroutines:")
				buf := make([]byte, 100000)
				stackSize := runtime.Stack(buf, true)
				t.Fatal(string(buf[:stackSize]))
			}
		}
		if p.inProgress != 0 {
			t.Errorf("unexpected query in progress")
		}
		if p.state != stopped {
			t.Errorf("unexpected proxy state, expected %d, got %d", stopped, p.state)
		}
	}

```

**Comment by miekg** [comment]
I think this should be fixed with #1738

**Comment by miekg** [comment]
Ha. Current master build:
~~~
018/04/27 20:54:42 [ERROR] plugin/file: Failed to open zone "miek.nl." in "go-test-tmpfile267335320": open go-test-tmpfile267335320: no such file or directory
ok  	github.com/coredns/coredns/plugin/file	3.383s
=== RUN   TestLess
--- PASS: TestLess (0.00s)
PASS
ok  	github.com/coredns/coredns/plugin/file/tree	1.011s
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
~~~

I need another deep dive into this - mostly like after kubecon

**Comment by miekg** [comment]
latest attempt #1754

**Comment by rdrozhdzh** [comment]
this may have the same reason as #1785, i.e. hang in NewServer().
Several factors leaded to coming out the problem
1. TestProxyClose() created a lot of tcp and udp connections, which then got to connection cache 
2. connection manager doesn't close cached connections on stopping (should be fixed in #1782)
So, when the next test in forward package tried to create new dnstest.Server it failed with quite high probability, because a lot of udp and tcp ports were still in use, and it was lower probability that the same port is free for both tcp and udp protocols.

**Comment by miekg** [comment]
can be closed again.
